interpreter/shared/src/main/scala/sigmastate/eval/CostingDataContext.scala

```diff
@@ -149,31 +149,36 @@ case class CSigmaProp(sigmaTree: SigmaBoolean) extends SigmaProp with WrapperOf[
   override def toString: String = s"SigmaProp(${wrappedValue.showToString})"
 }
 
-/** Implementation of the [[special.sigma.AvlTreeVerifier]] trait based on
+/** Implements operations of AVL tree verifier based on
   * [[scorex.crypto.authds.avltree.batch.BatchAVLVerifier]].
   *
-  * @see BatchAVLVerifier, AvlTreeVerifier
+  * @see BatchAVLVerifier
   */
-class CAvlTreeVerifier(startingDigest: ADDigest,
+class AvlTreeVerifier(startingDigest: ADDigest,
                        proof: SerializedAdProof,
                        override val keyLength: Int,
                        override val valueLengthOpt: Option[Int])
     extends BatchAVLVerifier[Digest32, Blake2b256.type](
-      startingDigest, proof, keyLength, valueLengthOpt)
-        with AvlTreeVerifier {
-  override def treeHeight: Int = rootNodeHeight
+      startingDigest, proof, keyLength, valueLengthOpt) {
+  def treeHeight: Int = rootNodeHeight
 
   /** Override default logging which outputs stack trace to the console. */
   override protected def logError(t: Throwable): Unit = {}
 }
+object AvlTreeVerifier {
+  def apply(tree: AvlTree, proof: Coll[Byte]): AvlTreeVerifier = {
```

Comment:
turned constructor of AvlTreeVerifier into private, added ScalaDoc for apply method, which serves as constructor (this is standard pattern in Scala, not sure what you mean by "not the best name"). 
It is created simply as `AvlTreeVerifier(tree, proof)`.

Reply:
I suppose that apply() is used around Scala ecosystem to create an object in consistent state. Which is not the case here - the state would be consistent only for lookup proofs. 

`createVerifier` was a dirty hack already, but now you are going further with it by endorsing a method used everywhere to create an object with inconsistent state mostly.

Instead of hiding dangerous behavior silently under apply(), you need to limit access to dirty code and provide a warning(better, via naming also), or do a builder.

Reply:
Please explain how createVerifier is a dirty hack?
Give an example of when "object with inconsistent state mostly" is created?
I don't understand you hints, please clarify.

Note, for each AvlTree operation, a new AvlTreeVerifier is created and not reused across operations. The role of createVerifier method is to do costing of creation itself.

Reply:
> you are expecting from resulting tree to have digest AFTER proof application

as far as I understand, after Verifier construction, the digest indeed corespoinds  to the digest of the tree reconstructed from the proof.
At least what I can learn from `require(startingDigest startsWith root.label)` in the computation of reconstructedTree field (which is done as part of constructor see topNode)

> resulting tree has digest BEFORE the proof

this is wrong

Reply:
AvlTree.digest is always connected with the proof in such a way, that when the verifier reconstructs the tree from the proof, the root node should have the avlTree.digest as the label.
This is checked in the Verifier constructor.
With this in mind I don't see any hacks you mentioned.

Reply:
> What check in the constructor are you talking about ?

https://github.com/input-output-hk/scrypto/blob/d925f9357d3a3f2ee2b3f1be5d7b82f1eb8f8b54/shared/src/main/scala/scorex/crypto/authds/avltree/batch/BatchAVLVerifier.scala#L224

Reply:
> > you are expecting from resulting tree to have digest AFTER proof application
> 
> as far as I understand, after Verifier construction, the digest indeed corespoinds to the digest of the tree reconstructed from the proof. At least what I can learn from `require(startingDigest startsWith root.label)` in the computation of reconstructedTree field (which is done as part of constructor see topNode)
> 
> > resulting tree has digest BEFORE the proof
> 
> this is wrong

Starting digest is indeed corresponding to the proof's tree. However, my point was not about that, rather, about e relation between tree and proof. 

I think as a quick workaround you can  call it `treeBeforeProof`, and update scaladoc accordingly.

Reply:
But, it is not tree before proof, my point is that it is actually the tree AFTER proof.
Clarified this point in the ScalaDoc.